python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,351 @@
+            size_loss = self.coff_budget * F.relu(
+                torch.sum(edge_mask) - self.coff_budget
+            )
+
+        # entropy
+        scale = 0.99
+        edge_mask = self.edge_mask * (2 * scale - 1.0) + (1.0 - scale)
+        mask_ent = -edge_mask * torch.log(edge_mask) - (
+            1 - edge_mask
+        ) * torch.log(1 - edge_mask)
+        mask_ent_loss = self.coff_connect * torch.mean(mask_ent)
+
+        loss = pred_loss + size_loss + mask_ent_loss
+
+        return loss
+
+    def concrete_sample(self, log_alpha, beta=1.0, training=True):
+        r"""Sample from the instantiation of concrete distribution when training.
+
+        Parameters
+        ----------
+        log_alpha : Tensor
+            A tensor representing the log of the prior probability of activating the gate.
+        beta : float, optional
+            Controls the degree of randomness in the gate's output.
+        training : bool, optional
+            Indicates whether the gate is being used during training or evaluation.
+
+        Returns
+        -------
+        Tensor
+            If training is set to True, the output is a tensor of probabilities that
+            represent the probability of activating the gate for each input element.
+            If training is set to False, the output is also a tensor of probabilities,
+            but they are determined solely by the log_alpha values, without adding any
+            random noise.
+        """
+        if training:
+            bias = self.sample_bias
+            random_noise = torch.rand(log_alpha.size()).to(log_alpha.device)
+            random_noise = bias + (1 - 2 * bias) * random_noise
+            gate_inputs = torch.log(random_noise) - torch.log(
+                1.0 - random_noise
+            )
+            gate_inputs = (gate_inputs + log_alpha) / beta
+            gate_inputs = torch.sigmoid(gate_inputs)
+        else:
+            gate_inputs = torch.sigmoid(log_alpha)
+
+        return gate_inputs
+
+    def train_step(self, graph, feat, tmp):
+        r"""Training the explanation network by gradient descent(GD)
+        using Adam optimizer
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            A homogeneous graph.
+        feat : Tensor
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        tmp : float
+            The temperature parameter fed to the sampling procedure.
+
+        Returns
+        -------
+        Tensor
+            A scalar tensor representing the loss.
+        """
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        pred = self.model(graph, feat).argmax(-1).data
+
+        prob, edge_mask = self.explain_graph(
+            graph, feat, tmp=tmp, training=True
+        )
+
+        self.edge_mask = edge_mask
+
+        loss_tmp = self.loss(prob.unsqueeze(dim=0), pred)
```

Comment:
Why do we need `prob.unsqueeze(dim=0)` here?

Reply:
Removed.